src/com/google/enterprise/adaptor/CommandStreamParser.java

```diff
@@ -467,6 +467,19 @@ public void readFromRetriever(DocId docId, Response response) throws IOException
     command = readCommand();
     while (command != null) {
       switch (command.getOperation()) {
+        case CONTENT:
+        case UP_TO_DATE:
+        case NOT_FOUND:
+          // Finish by putting accumulated ACL into response.
+          if (sendAclWithDocument) {
+            aclBuilder.setPermits(permits);
+            aclBuilder.setDenies(denies);
+            response.setAcl(aclBuilder.build());
+          }
+          break;
+        default:
+      }
+      switch (command.getOperation()) {
```

Comment:
Sort of. There's no "end of ACL" tag in the input, so the code was waiting until the end of the input (that is, the command loop) to process the ACL. But it's too late to call setAcl on the Response at that point, because respondXXX or getOutputStream have already been called. So I moved this block to the top of the loop, and check to see if we're about to process one of the must-be-last methods, and use that as a trigger to make sure that setAcl is called while it still can be.